packages/babel-parser/src/tokenizer/types.js

```diff
@@ -9,10 +9,9 @@
 // All token type variables start with an underscore, to make them
 // easy to recognize.
 
-// The `beforeExpr` property is used to disambiguate between regular
-// expressions and divisions. It is set on all token types that can
-// be followed by an expression (thus, a slash after them would be a
-// regular expression).
+// The `beforeExpr` property is used to disambiguate between 1) binary
+// expression (<) and JSX Tag start (<name>); 2) object literal and JSX
+// texts. It is set on the `updateContext` function in the JSX plugin.
```

Comment:
We can probably further simplify the `exprAllowed` logic in `updateContext` of JSX plugin, based on the usage here. However I would like to land it separately before this PR is too big for review.